pkg/addonmanager/controllers/agentdeploy/hookcontroller.go

```diff
@@ -0,0 +1,301 @@
+		return nil
+	}
+
+	// apply hookWork when addon is deleting
+	hookWork.OwnerReferences = []metav1.OwnerReference{*owner}
+	hookWork, err = c.applyWork(ctx, hookWork)
+	if err != nil {
+		meta.SetStatusCondition(&managedClusterAddonCopy.Status.Conditions, metav1.Condition{
+			Type:    "HookManifestCompleted",
+			Status:  metav1.ConditionFalse,
+			Reason:  "ManifestWorkApplyFailed",
+			Message: fmt.Sprintf("failed to apply manifestwork %v: %v", hookWork.Name, err),
+		})
+		c.addonClient.AddonV1alpha1().ManagedClusterAddOns(managedClusterAddonCopy.Namespace).UpdateStatus(
+			ctx, managedClusterAddonCopy, metav1.UpdateOptions{})
+		return err
+	}
+
+	meta.SetStatusCondition(&managedClusterAddonCopy.Status.Conditions, metav1.Condition{
+		Type:    "HookManifestCompleted",
+		Status:  metav1.ConditionFalse,
+		Reason:  "HookManifestIsNotCompleted",
+		Message: fmt.Sprintf("hook manifestWork %v is not completed.", hookWork.Name),
+	})
+
+	completed := hookWorkIsCompleted(hookWork)
+	if completed {
+		meta.SetStatusCondition(&managedClusterAddonCopy.Status.Conditions, metav1.Condition{
+			Type:    "HookManifestCompleted",
+			Status:  metav1.ConditionTrue,
+			Reason:  "HookManifestIsCompleted",
+			Message: fmt.Sprintf("hook manifestWork %v is completed.", hookWork.Name),
+		})
+	}
+
+	if !equality.Semantic.DeepEqual(managedClusterAddonCopy.Status, managedClusterAddon.Status) {
+		c.addonClient.AddonV1alpha1().ManagedClusterAddOns(managedClusterAddonCopy.Namespace).UpdateStatus(
+			ctx, managedClusterAddonCopy, metav1.UpdateOptions{})
+	}
+
+	if completed {
+		finalizer := removeFinalizer(managedClusterAddonCopy.Finalizers, constants.PreDeleteHookFinalizer)
+		managedClusterAddonCopy.SetFinalizers(finalizer)
+		_, err = c.addonClient.AddonV1alpha1().ManagedClusterAddOns(managedClusterAddonCopy.Namespace).Update(
+			ctx, managedClusterAddonCopy, metav1.UpdateOptions{})
+		return err
+	}
+
+	return nil
+}
+
+func (c *addonHookDeployController) applyWork(ctx context.Context, required *workapiv1.ManifestWork) (*workapiv1.ManifestWork, error) {
+	existingWork, err := c.workLister.ManifestWorks(required.Namespace).Get(required.Name)
+	existingWork = existingWork.DeepCopy()
+	if errors.IsNotFound(err) {
+		existingWork, err = c.workClient.WorkV1().ManifestWorks(required.Namespace).Create(ctx, required, metav1.CreateOptions{})
+		if err == nil {
+			c.eventRecorder.Eventf("ManifestWorkCreated", "Created %s/%s because it was missing", required.Namespace, required.Name)
+			c.cache.updateCache(required, existingWork)
+			return existingWork, nil
+		}
+		c.eventRecorder.Warningf("ManifestWorkCreateFailed", "Failed to create ManifestWork %s/%s: %v", required.Namespace, required.Name, err)
+		return nil, err
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	if c.cache.safeToSkipApply(required, existingWork) {
+		return existingWork, nil
+	}
+
+	if manifestsEqual(required.Spec.Workload.Manifests, existingWork.Spec.Workload.Manifests) {
```

Comment:
this will be good enough, since you still have manifestConfigs set.

Reply:
changed to compare the whole spec in the common apply func.